### DISCUSSION TITLE
How to set default base address for custom elf architecture?

### DISCUSSION THREAD

**Discussion by BoredPerson** [Vector35/binaryninja-api#2460] [Architectures] [answered]
Hey, I'm implementing an Architecture plugin for Binary Ninja. The binaries are standard elfs so I don't want to implement my own BinaryView. I registered my architecture with the elf BinaryViewType:
`BinaryViewType["ELF"].register_arch(247, Endianness.LittleEndian, Architecture["EBPF"])`

I tried to set an callback which then calls BinaryView.rebase but that causes Binary Ninja to freeze.
```
def bv_callback(bv: BinaryView):
      if bv.arch.name == "EBPF" and not REBASED:
              bv = bv.rebase(MM_PROGRAM_START, force=True)

BinaryViewType["ELF"].add_binaryview_initial_analysis_completion_event(bv_callback)
```

**Comment by rssor** [comment, marked as answer]
So there's two ways to go about doing this, presently:

It should be possible to set this during load using load options to set the base address manually (see: "Image Base Address" option).

Another more involved option would be to use a derived BinaryView. Essentially, you'd define a new BinaryViewType, but in the initialization method you'd call and create a normal ELF binaryview, change the segments/sections around to reflect what you want, then return that ELF view as if it was your own view. This would also require you to munge the ELF header in some way to make sure the file gets diverted to your own view instead. 

The issue with your current approach is just that the callback there isn't able to safely interact with the BinaryView object in that way.

Longer term, this sort of thing should be helped along significantly by work such as https://github.com/Vector35/binaryninja-api/issues/1977 which is being worked on presently.
